### DISCUSSION TITLE
GoPro as dashcam : Issue on Motion detection trigger in a scheduled interval

### DISCUSSION THREAD

**Discussion by esiremalaysia** [gopro/labs#322] [General]
### Use case :
Trying to use my gopro as a dashcam while a motion detection occurs in a specific scheduled interval using 'boot' command. Note that the gopro is continuously powered by a powerbank and cooled with a fan (in case of hot weather).

### Labs documentation :
I find the actual documentation for 'boot' command here (https://gopro.github.io/labs/control/extensions/), it says  : e.g. !10:00NoMBOOT=”!SM3D4H5!R” <- Wake at 10am, then motion capture continuously. But the instructions does not work on the hero10 (also tried on the hero8 and do not work neither). 

### Troubleshooting :
Here, the command I was troubleshooting with : 

**CASE 1** - Start Motion Detection continuously: !SM4H2!R (✅ working)
**CASE 2** - Start Motion Detection + Hindsight (15s window) continuously: mVhS1!SM4H2hS0!R (✅ working and already solved #187)
**CASE 3** - Start Motion Detection continuously on a scheduled time at 8:00 AM : !08:00NoMBOOT="!SM4H2!R" or !08:00NoMBOOT=!SM4H2!R (❌ failed) -> boot at scheduled time OK, but the motion detection FAILED with message error*
**CASE 4** - Start Motion Detection continuously on a scheduled time at 8:00 AM + Hindsight: !08:00NoMBOOT="mVhS1!SM4H2hS0!R" (❌ failed) -> boot at scheduled time OK, Hindsight OK, but the motion detection FAILED with message error*
**CASE 5** - Boot + Motion detection + Adding an SD card based boot file (base on this example : https://gopro.github.io/labs/control/imutrigger/)  : !MQRDR=1!08:00N!MBOOT="!Lmotion"!SAVEmotion=!SM4H2!R (❌ failed) -> same error from case 3 + but when restarting manually the gopro the motion detection begin without taking into account the time scheduled condition
**CASE 6** - Start Motion Detection continuously on a scheduled time at 8:00 AM and stop it at 20:00 AM : help❓
**CASE 7 (final)** - All the above (motion + hindsight + start/stop schedule ) + upload on the cloud : help❓

### Issue (MAIN) :
*Message error each time using 'boot' command:  "CANCEL MOTION TRIGGER"
![gopro_motion_issue](https://user-images.githubusercontent.com/42617463/234718052-2ebd90cb-b473-494c-9c67-a64e805f9ab1.jpeg)

It is kinda frustrating to see that every single step can work separately on the hero10 and hero8 but when we combine with boot commands everything failed.

### Issue (SECONDARY) : 
I want the rear LCD screen to be turned OFF on hero10 while processing the script above because it :
- drains too much battery for nothing
- waste of resources
- heats the gopro 
- that's not unobtrusive (especially if used in a car at night/bad weather)

I saw the issue here #77, but unfortunately technical support @dnewman-gpsw says that's not possible due to the architectural difference. I feel really sad that it works on previous models hero8, hero9 and newer model hero11 but not on hero10. Feel a little scammed on that issue because as a gopro customer and a developer every working feature from previous model must be included on newer model, especially with cameras industry. It might be a solution, if you can reopen that issue I will highly appreciate. 

### Logs (for case 4) :
```
[2023-04-26 18:41:08] Labs:Main Start 
[2023-04-26 18:41:08] Labs:Initialize 
[2023-04-26 18:42:48] Read QR:!18:44NQv1oMBOOT="mVhS1!SM4H2hS0!R" 
[2023-04-26 18:42:52] Shutdown and Sleep:66 seconds 
[2023-04-26 18:42:53] shutdown:rtc wake scheduled 
[2023-04-26 18:44:01] Woke with Alarm:[18:44m01]!NQv1oMBOOT="mVhS1!SM4H2hS0!R" 
[2023-04-26 18:44:01] Labs:Main Start 
[2023-04-26 18:44:01] Labs:Initialize 
[2023-04-26 18:44:04] Woke with Command:[18:44m01]!NQv1oMBOOT="mVhS1!SM4H2hS0!R" 
[2023-04-26 18:44:04] Start Now:Past by 4 seconds 
[2023-04-26 18:44:04] Countdown:complete 
[2023-04-26 18:44:04] Timer wake command:!NQv1oMBOOT="mVhS1!SM4H2hS0!R" 
[2023-04-26 18:44:04] Labs:QR Detection Waking 
[2023-04-26 18:44:04] Labs:QR Detection Ready 
[2023-04-26 18:44:07] Boot command:mVhS1!SM4H2hS0!R 
[2023-04-26 18:44:07] Command order:mVhS0hS1!SM4H2hS0!R 

```
_Running on GoPro Hero 10, Firmware 01.50.70_

### Goal :
My goal is at least to find a solution to make **CASE 3** functional as written in the documentation which is my biggest concern.

I take time to meticulously point the issue and improve gopro labs features, and expect a solution from it. Thanks in advance for the support.

**Comment by dnewman-gpsw** [comment]
As for "Feel a little scammed on that issue because as a gopro customer and a developer every working feature from previous model must be included on newer model, especially with cameras industry. "  That doesn't not apply to Labs and it would kill Labs if it did.  Not all features from old Labs are supported on newer cameras, as the engineering for Labs needs to be nibble. We've had experimental features that don't work on the next model. e.g. encryption, introduced on HERO9, didn't work in HERO10 or 11.  

I expect HERO10 will see another Labs update this year, not anytime soon.  BUT, if you read this forums close enough, patch versions are sometimes shared (not QA'd, not official,) I have not idea if the latest patch address HERO10 sleep issue so try it, good for few days more: https://www.dropbox.com/transfer/AAAAAFlTQ8vpofFqfchhxV5MlkWijDyyn9PoChwnQKAlQUV4OEtvzOw

**Reply by esiremalaysia** [reply]
Thanks for your support, now **CASE 3** (motion detection at boot up : _!08:00NoMBOOT="!SM4H2!R_) is working fine. 

But **CASE 2** (motion detection + hindsight : _mVhS1!SM4H2hS0!R_) is not working anymore. Only Hindsight was activated and no screen message this time for motion detection. 

Did I forget something on the script ? Cauz on the previous labs firmware (1.50.71) it worked, not anymore with (1.50.72).

Thanks, we are close to the goal now :)

**Reply by dnewman-gpsw** [reply]
Case is broken on H11 too, that H10 build is using HERO11 Labs logic. However, you will find a lot of the new scripting works on that build. So maybe it is a win.

**Reply by esiremalaysia** [reply]
Could you give me more detail in why combined actions work on 1.50.71 and not on 1.50.72 ? It's like sacrificing a working feature for another. 

I'm afraid that if adding Hindsight to the motion detection fail, this will have an impact on case 5, 6, 7 because it requires even more conditions/actions. Trying to understand as it is not open-source, I cannot do anything on my side.

**Reply by dnewman-gpsw** [reply]
I think is only the hindsight is prerecording, that looks like recording in most ways, so it is impacting motion detection starts.  Hindsight is was weird corner-case mode that Labs struggles with.  It is not the combination of modes, it only hindsight.

**Reply by esiremalaysia** [reply]
Yes I can understand that H10 build is using H11 Labs logic. But what can be enhanced is to take what work on 1.50.71 and put it back into 1.50.72. 

I redo all the whole manipulation with firmware `1.50.71` and `1.50.72` for the following command so we can understand through logs what happens precisely : `mVhS5!SM4H2hS0!R`

What I do : 
1) Run same command on both version after hindsight stabilize for 5s
2) Make a movement to activate motion detection one time only
3) Stop immediately the camera just after

## GP10 firmware 1.50.71 : 

### Logs 
```
[2023-04-27 22:04:00] Update:01.50.70 to 01.50.71 
[2023-04-27 22:04:08] Update:Approved 
[2023-04-27 22:04:10] Labs:Main Start 
[2023-04-27 22:04:13] Labs:Main Start 
[2023-04-27 22:04:13] Labs:Initialize 

[2023-04-27 22:04:26] Read QR:!MDBGL=1 
[2023-04-27 22:04:26] Use CMD:!MDBGL=1 
[2023-04-27 22:04:26] ACTION:!MDBGL=1

[2023-04-27 22:04:37] Read QR:mVhS5!SM4H2hS0!R 
[2023-04-27 22:04:37] Use CMD:mVhS5!SM4H2hS0!R 
[2023-04-27 22:04:38] token:V 
[2023-04-27 22:04:38] CAMMODE token:V 
[2023-04-27 22:04:38] token:S5 
[2023-04-27 22:04:38] HANDLE_MASK token:S5 
[2023-04-27 22:04:38] presets_edit:done 
[2023-04-27 22:04:38] ACTION:!SM4H2hS0!R 
[2023-04-27 22:04:41] GP2 Temp.:49 
[2023-04-27 22:04:46] GP2 Temp.:52 
[2023-04-27 22:04:53] token:S0 
[2023-04-27 22:04:53] HANDLE_MASK token:S0 
[2023-04-27 22:04:53] presets_edit:done 
[2023-04-27 22:04:53] ACTION:!R 
[2023-04-27 22:04:53] Labs:QR Detection Waking 
[2023-04-27 22:04:54] Labs:QR Detection Ready 
[2023-04-27 22:04:54] token:V 
[2023-04-27 22:04:54] CAMMODE token:V 
[2023-04-27 22:04:54] token:S5 
[2023-04-27 22:04:54] HANDLE_MASK token:S5 
[2023-04-27 22:04:55] presets_edit:done 
[2023-04-27 22:04:55] ACTION:!SM4H2hS0!R 
[2023-04-27 22:04:57] GP2 Temp.:50 
[2023-04-27 22:04:57] Labs:QR Detection Ready 
[2023-04-27 22:05:01] GP2 Temp.:53 
[2023-04-27 22:05:08] GP2 Temp.:51 
```

### Result
✅ Success -> Both motion detection and Hindsight work continuously  
<img width="903" alt="V15071_demo" src="https://user-images.githubusercontent.com/42617463/234994045-e2458ee4-dc3d-4a9d-bdf9-9deb53f29111.png">


## GP10 firmware 1.50.72 : 

### Logs 
```
[2023-04-27 22:11:29] Update:01.50.71 to 01.50.72 
[2023-04-27 22:11:37] Update:Approved 
[2023-04-27 22:12:07] Labs:Main Start 
[2023-04-27 22:12:20] Labs:Main Start 
[2023-04-27 22:12:20] Labs:Initialize

[2023-04-27 22:12:25] Read QR:!MDBGL=1

[2023-04-27 22:12:29] Read QR:mVhS5!SM4H2hS0!R 
[2023-04-27 22:12:30] Use CMD:mVhS5!SM4H2hS0!R 
[2023-04-27 22:12:30] Command order:mVhS0hS5!SM4H2hS0!R 
[2023-04-27 22:12:31] token:V 
[2023-04-27 22:12:31] CAMMODE token:V 
[2023-04-27 22:12:31] token:S0 
[2023-04-27 22:12:31] HANDLE_MASK token:S0 
[2023-04-27 22:12:31] token:S5 
[2023-04-27 22:12:31] HANDLE_MASK token:S5 
[2023-04-27 22:12:31] presets_edit:done 
[2023-04-27 22:12:31] ACTION:!SM4H2hS0!R 
[2023-04-27 22:12:31] capture started:user input 
[2023-04-27 22:12:33] GP2 Temp.:44 
[2023-04-27 22:12:37] GP2 Temp.:47
```

Seeing a strange notation here, to confirm : [2023-04-27 22:12:30] Command order:mV**hS0**hS5!SM4H2hS0!R 

### Result
❌ Failed -> Only Hindsight executed, seems that motion detection was ignored. 
<img width="1026" alt="V15072_demo" src="https://user-images.githubusercontent.com/42617463/234994426-2742fc02-aead-4ae2-9348-1cdba2c9c970.png">

If these elements can help to fix the issue, it will be a great step for gopro into dashcam industry !

Thanks in advance

**Reply by dnewman-gpsw** [reply]
"Command order:mVhS0hS5!SM4H2hS0!R" this is correcting for something experienced in H11,  users where adding hS commands after !S, causing a problem, this code moving that hS0 forward, even though it is not need for you code. Hindsight is weird with Labs. And there is currently a bug, nothing you can do at the moment.

**Reply by esiremalaysia** [reply]
Ok then will wait to a bug fix for H10/H11 then, hope it will be soon. Fingers crossed ! 

In the meanwhile, I have two request : 
- Can you explain me why rear LCD screen cannot be turned OFF on H10 while processing motion script, is it possible to fix that on the future release ? As H10 build is using H11 Labs logic now, maybe there is a way to make that possible right ?
- Can you drop here the latest un-official firmware release for the H8 ? The latest official firmware that I found for the H8 is `v2.51.75, July 2022`. There is the same sleep issue as of the H10 on `v1.50.71`, so maybe there is a fix in a non-official version so that I can try.

PS : I will continue to work on my scripts to make **case  5, 6, 7** possible, so that I can update this feature sheet. As soon as the Hindsight bug will be fixed, will update the **case 2 and 4** to make every dashcam use-cases functional.

Thanks in advance David Newman

**Reply by dnewman-gpsw** [reply]
Making some progress on the hindsight issue.

1) Motion Detection needs to run while recording, the fastest buffer available on that model (pre-scaled for speed) was the one for the screen. Screen off, no data for motion detection. Could it use the newer H11 techniques? Not sure, if they where developed for H10 (through 1.50) then maybe. Labs never moves production code from one camera model to other--Labs tries not to brick cameras. The Labs core, command processing, is the only code that remains somewhat portable.

2) There is none. July 2022 is a pretty recent build for that camera, I expect that will be the last.

**Reply by esiremalaysia** [reply]
Hello, 

I'am trying to create a script for case **CASE 6 (= Start Motion Detection continuously on a scheduled time at 8:00 AM and stop it at 20:00)** but this won't end as programmed. Here the script : `!08:00NoMBOOT="!SM4H2!20:00E!R` or even `!08:00NoMBOOT="!SM4H2!R!20:00E`

According to the documentation (https://gopro.github.io/labs/control/actions/) : !timeE must end at time if in hh:mm form, or after n seconds. But even trying `!10E`, `!60E` or `!20:00E` does not work, why ?

**Reply by dnewman-gpsw** [reply]
!E command can only be paired with !S command, not !SM or !SI etc.  This one of the many reason to obsolete the older triggers for the new scripting. !SM is both a start and end trigger, so script processing doesn't see !20:00E until the capture has ended due to the lack motion.  It is not possible if motion is continuous.  The closest would be something like !08:00NoMBOOT="!SM4H2<20:00!R"

**Reply by esiremalaysia** [reply]
Thanks David, I figure it out to have a stable script that makes possible simultaneously **CASE 1, 3, 5, 6** (all case appart Hindsight which are **CASE 2, 4, 7**). 

Here the working script : `!08:00N!MBOOT="!Lmotion"!SAVEmotion="!08:00N!MBOOT="<20:00!SM4H2~!10O!R""`
What it does : Start Motion Detection continuously on a scheduled time at 8:00 AM and stop it at 20:00 AM, the script named 'Lmotion' is activated each day because it is stored on the SD card. After 20:00, the detection end, and after 10 seconds the camera shutdowns. And the cycles restart (have to confirm in the future as it was a test, not in a real case scenario).

1. My main question is why this script which are normally the one that must work don't work properly : `!08:00N!MBOOT="!Lmotion"!SAVEmotion="<20:00!SM4H2~!10O!R"`,  after 20:00 the motion detection continues and don't turn off, why ? 

2. My second big concerns is when does Hindsight will be fixed as well for H10 and H11, any future experimental release (unofficial) ?

**Reply by dnewman-gpsw** [reply]
1. You made creative but weird commands: !08:00N!MBOOT="!Lmotion"!SAVEmotion="<20:00!SM4H2~!10O!R"

This says wait until 8am, then save create the boot script. Weird. Remove the first !08:00N. The is the script saved is odd.  <20:00!SM4H2~!10O!R  This is if less than 8pm, motion detect else shutdown (after sleeping the camera for 10s), loop.  This is nothing to wake the camera at 8am.   It should be <20:00!SM4H2~!08:00R!R

2.  Other bugs to fix before any releases official or not.

**Reply by esiremalaysia** [reply]
Your script is not working after trying for one week... : `<20:00!SM4H2~!08:00R!R`

It just activate the motion detection if <20:00 and then sleep till 8:00, and then after waking up it does literally nothing (no more motion detection) 

What I want is a script similar to that so it will save on the SD card : `!08:00N!MBOOT="!Lmotion"!SAVEmotion="<20:00!SM4H2~!10O!R"` but this script does not work as said below

**Reply by esiremalaysia** [reply]
@dnewman-gpsw up !

Assistance may be appreciated

**Comment by esiremalaysia** [comment]
Hello @dnewman-gpsw,

I recently tested the new H12 for the **motion detection + hindsight (15s window) continuously**, and I find out that this QR code previously working on H10 and H11 : `mVhS1!SM4H2hS0!R` is not working on H12. Another discussions also solved the issue for H11 here #187. 

Here the logs : 

```
[2023-09-25 19:25:47] Read QR:!MDBGL=1  
[2023-09-25 19:25:47] unknown token:

[2023-09-25 19:25:49] Read QR:mVhS1!SM4H2hS0!R 
[2023-09-25 19:25:50] Use CMD:mVhS1!SM4H2hS0!R 
[2023-09-25 19:25:50] token:V 
[2023-09-25 19:25:51] CAMMODE token:V 
[2023-09-25 19:25:51] token:S1 
[2023-09-25 19:25:51] HANDLE or HS token:S1 
[2023-09-25 19:25:51] presets_edit:done 
[2023-09-25 19:25:51] ACTION:!SM4H2hS0!R 
[2023-09-25 19:25:51] display 250ms:start 2 seconds 
[2023-09-25 19:25:51] display 250ms:start 2 seconds 
[2023-09-25 19:25:51] display 250ms:start 2 seconds 
[2023-09-25 19:25:52] display 250ms:start 1 seconds 
[2023-09-25 19:25:53] STARTCAPTURE:7 
[2023-09-25 19:25:53] STARTCAPTURE:7 
[2023-09-25 19:25:54] GP2 Temp.:40 
[2023-09-25 19:25:58] STOPCAPTURE:22 
[2023-09-25 19:25:58] token:S0 
[2023-09-25 19:25:58] HANDLE or HS token:S0 
[2023-09-25 19:25:58] presets_edit:done 
[2023-09-25 19:25:58] ACTION:!R 
[2023-09-25 19:25:59] GP2 Temp.:43 
[2023-09-25 19:26:00] token:V 
[2023-09-25 19:26:00] CAMMODE token:V 
[2023-09-25 19:26:00] token:S1 
[2023-09-25 19:26:00] HANDLE or HS token:S1 
[2023-09-25 19:26:02] presets_edit:done 
[2023-09-25 19:26:02] ACTION:!SM4H2hS0!R 
[2023-09-25 19:26:03] GP2 Temp.:40 
[2023-09-25 19:26:06] STARTCAPTURE:32 
[2023-09-25 19:26:06] STARTCAPTURE:32 
[2023-09-25 19:26:09] GP2 Temp.:42 
[2023-09-25 19:26:15] STOPCAPTURE:59 
[2023-09-25 19:26:15] token:S0 
[2023-09-25 19:26:15] HANDLE or HS token:S0 
[2023-09-25 19:26:15] presets_edit:done 
[2023-09-25 19:26:15] ACTION:!R 
[2023-09-25 19:26:15] GP2 Temp.:44 
[2023-09-25 19:26:17] token:V 
[2023-09-25 19:26:17] CAMMODE token:V 
[2023-09-25 19:26:17] token:S1 
[2023-09-25 19:26:17] HANDLE or HS token:S1 
[2023-09-25 19:26:17] presets_edit:done 
[2023-09-25 19:26:17] ACTION:!SM4H2hS0!R 
[2023-09-25 19:26:21] GP2 Temp.:42 
[2023-09-25 19:26:24] STARTCAPTURE:78 
[2023-09-25 19:26:24] STARTCAPTURE:78 
[2023-09-25 19:26:25] GP2 Temp.:45 
[2023-09-25 19:26:28] STOPCAPTURE:99 
[2023-09-25 19:26:29] token:S0 
[2023-09-25 19:26:29] HANDLE or HS token:S0 
[2023-09-25 19:26:29] presets_edit:done 
[2023-09-25 19:26:29] ACTION:!R 
[2023-09-25 19:26:30] token:V 
[2023-09-25 19:26:31] CAMMODE token:V 
[2023-09-25 19:26:31] token:S1 
[2023-09-25 19:26:31] HANDLE or HS token:S1 
[2023-09-25 19:26:33] presets_edit:done 
[2023-09-25 19:26:33] ACTION:!SM4H2hS0!R 
[2023-09-25 19:26:33] GP2 Temp.:42 
[2023-09-25 19:26:33] capture started:user input 
```

So what happens is that the gopro do the following : **motion detection + hindsight** for 3 times perfectly, but then for the fourth time it stopped and the logs atm was `capture started:user input`. And the screen come back. to normal with the following text on top : `Labs Processing`. What does that mean ? And why ?

**Reply by dnewman-gpsw** [reply]
I run script, !SM4H2hS0!R, nine times without any failure, on HERO12.  "capture started:user input " should only happy if you start a capture with a shutter press, or via the remote or App, I can't see any other way that can happen.  

"Labs Processing" means there is a loop (!R), so you know it is running a script.  Sometime it stays on when canceling the script, which is a minor bug.

**Reply by esiremalaysia** [reply]
This script only involves `!SM4H2hS0!R` motion detection only, and I myself test it and it is working indefinitely.
My script involving motion detection + hindsight (15s window) continuously :  `mVhS1!SM4H2hS0!R`
And for the "capture started:user input" in the logs below, I didn't press any button, just turn off the camera to take the logs from the sdcard. As said this latest script seems to be working on h10 and h11 according this solved issue #187 

Thanks for the "Labs Processing" explanation, now I know.

Can you tell me if my script involving both MD and hindsight is working on your side ?

**Reply by dnewman-gpsw** [reply]
Sorry, yes I meant it works with hs1, the full command was used.  I do not get that illogical error.

**Reply by esiremalaysia** [reply]
Yes, will this bug will be corrected for H12 soon ? 

The issue is that, this command don't run indefinitely in the H12 : `mVhS1!SM4H2hS0!R`, it stops randomly at some moment sometimes it stop at the first loop, the third or more but it stops and does not respect the `!R` which is suppose to repeat the script for each motion detection trigger. 

Seems that this bug was corrected with an alternative Labs update on H10/H12 according to #187, maybe a workaround to do for H12 ? 

Thanks in advance

**Reply by dnewman-gpsw** [reply]
It is working on my HERO12s at the moment (running two continuously), however I did see the MD not start a couple of times.  I solved that with hS1!1N!SM4H2hS0!R just give the camera a second between initializing the Hindsight and starting the motion detection.

**Reply by esiremalaysia** [reply]
Interesting, I didn't even think about adding time between each process, now I know. I tested it and it is working on H12, thanks. Now I will work on the other features and keep all updated. Have a good day @dnewman-gpsw and thanks again.

**Comment by esiremalaysia** [comment]
## Introducing GoPro as a dashcam

### Camera compatibility : 
- [x]  _tested on H12, firmware v1.12.70, September 29, 2023_
- [x]  _tested on H11, firmware v2.12.70, August 2023_
- [x]  _tested on H11 mini, firmware v2.30.70, May 2023_

### Dashcam Mode while driving (day/night)
- **Use case explained and scripts :** The script boot automatically upon USB power, it will then wait until it detects acceleration before starting recording. The script will also create space as your SD card fills. It will remain recording until USB power is removed but it will continue to record while it detects any physical camera motion, but stopped 10 seconds after all motion has ceased. The camera will remain on for a further 20 seconds just in case USB power is returned, but otherwise it will shutdown, ready to automatically power up again and start all over. 
   - ✅ 1. All the above (only on H11) :`!MWAKE=2!MTUSB=1!MBOOT="!Ldcam"!SAVEdcam=<u0!R97!C16<a0.05!R9=L 1>L99!C16 =L0=Bt:B>u0<r0!S =AB>u0=AB>a0.05=AB=B-A>B10>r0<u0!EB30!O!R18!B2`
   - ✅ 2. Start on power USB with no motion detection at the end (H11/H12) : `!MBOOT="!Lbt"!SAVEbt="Dashcam"+>u0<r0!C16+!S+=A1>A1<u0>r0<m1=A+1+!1N~=A1>A6>r0!E+!1O!R10`
- _**Additional thing :** script is clever and detects that it has been started without USB power and behave like a regular GoPro. So you can use it as a dashcam, take it out, uses a regular GoPro without ever having to re-scan a QR code_ 
- related issue : #484, #396
- demo [here](https://www.youtube.com/watch?v=T87RekjaVA4)

### Dashcam Mode when parked (day)
   _As gopro cameras had bad low light performance until H12 (01/10/23), it is not interesting to configure it for night time, but if there is a good outdoor lighting, just skip intervals condition and directly use the QR code n°2_

- **All use cases explained and scripts (H11/H12) :** 
  - ✅ 1. Start Motion Detection continuously :  `!SM4H2!R`
  - ✅ 2. Start Motion Detection + Hindsight (15s window) continuously : `hS1!1N!SM4H2hS0!R`
  - ✅ 3. Start Motion Detection continuously on a scheduled time (from 8:00AM to 20:00AM) : `<20:00!SM4H2~!08:00R!R`
  - ✅ 4. Start Motion Detection continuously on a scheduled time (from 8:00AM to 20:00AM) + Hindsight (15s window) : `<20:00hS1+!1N+!SM4H2+hS0~!08:00N!R`
  - ❌ 5. Start Motion Detection continuously on a scheduled time **with a precise starting time** (from 8:00AM to 20:00AM) + Hindsight (15s window) : `>05:00<20:00hS1+!1N+!SM4H2+hS0~!08:00N!R`
  - 🔄 6. All the above (motion + hindsight + daylight schedule ) + upload on the cloud : todo

**note:** The advantage of script 5 over 4 is that, script 5 had more control on time loops and when script must be executed. For example, if you run the script at 3AM in the morning, script 4 will enter the dashcam script immediately and start processing due to his single condition loop whereas script 5 will shutdown and run the dashcam script only if the time is between 5AM and 20AM due to his stacked conditionals loop. Use case n°5 is still under stress testing, will keep updated.

### Logs debugging 
Improve logging verbosity (Debug Level 1)
- `!MDBGL=1`

### Reset Labs
- `!RESET`

**Reply by esiremalaysia** [reply]
@dnewman-gpsw one last question, why I couldn’t make conditionals specific intervals as described on the [documentation](https://gopro.github.io/labs/control/actions/), I found this example : `>09:15<10:00!S+”Hello World”+!60E~!09:30N!R` -> and it does not work on the H12.

So according to this exemple, I want this script to be operational but it does not even activate 
`>05:00<2:00!S+hS1+!1N+!SM4H2+hS0~!08:00N!R`

Is there another alternative way of coding to make the script operational on H12 ?

**Reply by dnewman-gpsw** [reply]
First uses the wrong quotes, only use "  not ”  (ASCII characters only.)
Second isn't using 24:00 time format.  >05:00<02:00 (not <2:00) 
Also you can't turn on hS1 after !S, then start motion det, while recording. This needs some work.

**Reply by esiremalaysia** [reply]
Thanks for the reactivity,

1) Okey I see, I just followed the documentation, which use `”` instead of `"`. Need to be corrected in the [actions documentation section](https://gopro.github.io/labs/control/actions/) with a mention of ASCII characters to avoid future undesired issue.
2) It fix the thing, here the corrected version : `>05:00<20:00+hS1+!1N+!SM4H2+hS0~!08:00N!R`

Update will be added on the feature sheet below.

**Reply by dnewman-gpsw** [reply]
1. Yep, that should be fixed.  Note: These pages are open source, you can contribute any correction as a pull request.
2. Is slightly wrong still: ```>05:00<20:00hS1+!1N+!SM4H2+hS0~!08:00N!R```

**Reply by esiremalaysia** [reply]
1. Sure will contribute on my free time
2. Yes thanks for the typo. Oke here the result after testing the script at different time of the day, I found that it does not work as expected.

### Problem
There is a bug in the **stacked conditionals based on time**, when this script `>05:00<20:00hS1+!1N+!SM4H2+hS0~!08:00N!R` is executed :
- before 05:00 AM : 
  - the camera is running doing nothing with the label "Labs processing"  ✅  --> Loop on the first IF
  - If the camera is on until 8:00 AM, it start dashcam mode ✅
  - after 20:00 AM, the motion detection stops ✅ the Hindsight mode stops ✅ the gopro go to sleep mode ✅
  - but when it wake up, it go back to sleep and re-program to wake up at 8:00 the next day. Infinite loop on going through sleep mode. ❌  --> normally it should go through the dashcam mode as it is 8:00, which is not happening. Why ?
 
- after 05:00 AM and before 20:00 AM :
  - the script go through the dashcam mode ✅
  - after 20:00 AM, the motion detection stops ✅ the Hindsight mode stops ✅ the gopro go to sleep mode ✅
  - the camera wake up and what ? 🔄 to test

- after 20:00 AM :
  - the script go to sleep and wake up at 8:00 ✅
  - but when it wake up, it go back to sleep and re-program to wake up at 8:00 the next day. Infinite loop on going through sleep mode. ❌  --> normally it should go through the dashcam mode as it is 8:00, which is not happening. Why ?

There is definitively a conditionals bad interpretation, i expect this script to do this :
```
if(current_time >= 5:00)
{
    if(current_time <= 20:00)
    {
       Hindsight
       Wait 1 second
       Motion detection
    }
    else
    {
       Sleep until 8:00 the next day
    }
}
Repeat
```



But it does not behave as expected, need help on that to fix it, can you please try this on your side, 

Thanks in advance

**Reply by dnewman-gpsw** [reply]
What did the log say?  The errors will be in there. 

> the camera is running doing nothing with the label "Labs processing" ❌ --> Normally it should go to sleep mode. Why ?

Why is this wrong? This is what script askes for.  The script is running but doing nothing, so it shows the "Labs processing" message.  The camera remains awake as there is an assumption if you running a script you want it to remain running as there are command to sleep the camera.

> after 20:00 AM, the motion detection stops ✅ but Hindsight still remain running ❌and do not go to sleep mode (to wake up the next day) ❌ --> Why ?

This might require another !1N.  The log will help explain.

I can't try all users 24 hour scripts, you will need provide the debug logs.

**Reply by esiremalaysia** [reply]
Yes my bad I just corrected the previous comment with the actual error, everything doing pretty good but when the script enter the sleep mode, it does not always wake up the next day. I found that, when it enter the sleep mode, sometimes the gopro turn back ON several times for no reason and go back to sleep, maybe is it to keep time in memory while he is asleep ? 

What will be ideal is to apply the else to the first condition like that, so the script will be complete (and it will avoid the gopro to stay ON if it is activated after 23:59) : 

```
if(current_time >= 5:00)
{
    if(current_time <= 20:00)
    {
       Hindsight
       Wait 1 second
       Motion detection
    }
    else
    {
       Sleep until 8:00 the next day
    }
} else {
       Sleep until 8:00 the next day
}
Repeat
```
But I read the documentation and it says the following : _"However the else can only be applied to the last condition."_
Meanwhile does all the "else" on the stacked conditioned can be manipulable in a future update (not only the last one) ? 

I will test the same script for 24 hours and provide more logs feedback.

**Reply by dnewman-gpsw** [reply]
Just do:

>05:00<20:00hS1+!1N+!SM4H2+hS0~!08:00N<05:00!08:00N!R

```
if(current_time >= 5:00)
{
    if(current_time < 20:00)
    {
       Hindsight
       Wait 1 second
       Motion detection
    }
    else
    {
       Sleep until 8:00 the next day
    }
} 
if(current_time < 5:00)
       Sleep until 8:00 same day
}
Repeat
```

**Reply by esiremalaysia** [reply]
Great idea. This script displays `QRcode non-GoPro` due to a typo error.

Corrected version : `>05:00<20:00hS1+!1N+!SM4H2+hS0~!08:00N<05:00!08:00N!R`

Will test it and keep updated !

**Comment by esiremalaysia** [comment]
Issue fixed 🥇 

After several tests and stress testing, the full needed features are finally working mainly on H12 and partially on H11,
Thanks @dnewman-gpsw 

Advanced details [here](https://github.com/esiremalaysia/gopro/issues/1)
